### DISCUSSION TITLE
How to enumerate all dynamic/theme resources?

### DISCUSSION THREAD

**Discussion by toomasz** [AvaloniaUI/Avalonia#14734] [General]
Hey,
I would like to create view where are brush resources for current theme can be shown (as it would be useful while developing controls/views for my app)

Question one: Is there a way to enumerate all current theme resources keys?
 - App.Resources dictionary only contains my resources from current app
 - control.Resources is empty - expected as haven't added any to my control
- app.TryGetResource - works but I need to know resource key

Side question:
 - If there is no way to do that - Is there a list of theme resources/colors that can be used in app/custom controls? For example - Let's say I'm creating some control with Border and would like it style it according to current theme

This is what I'm aiming to create (currently few resources from Fluentv2.axaml - goal is to enumerate them all)
![image](https://github.com/AvaloniaUI/Avalonia/assets/3065454/36adb5e2-60d8-4e40-bdc2-b1959151ac36)
![image](https://github.com/AvaloniaUI/Avalonia/assets/3065454/8ea00372-21a8-4628-978e-b4eeaff0b77e)

**Comment by timunie** [comment]
FluentAvalonia had such a page for their theme. Maybe you can check out how they do it.

**Reply by toomasz** [reply]
They only have few colors and they are manually added. Not what I'm looking for.
![image](https://github.com/AvaloniaUI/Avalonia/assets/3065454/d39c1a0f-9357-41cf-89fc-d6872be98599)

**Comment by maxkatz6** [comment]
> Question one: Is there a way to enumerate all current theme resources keys?

Recursively enumerate all target.Resources, target.Resources.MergedResources, target.Styles.Resources, target.Styles.MergedResources (+nested styles if any).
There is no convenient API for that, as it's not really a common need.

> If there is no way to do that - Is there a list of theme resources/colors that can be used in app/custom controls? For example - Let's say I'm creating some control with Border and would like it style it according to current theme

It's very theme dependent. I.e. we have Fluent and Simple themes built in, but Material.Avalonia might do things differently.
Specifically, about FluentTheme, we have these base colors https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Accents/BaseColorsPalette.xaml, which can be redefined by the used like this: https://docs.avaloniaui.net/docs/basics/user-interface/styling/themes/fluent#creating-custom-color-palettes-with-online-editor
